### PR TITLE
Write all deals from special feeds into a file

### PR DESCRIPTION
**Pull request by sulaimonlasisi**
This branch gets all the deals from special feeds like clearance, preorder and specialbuy.
It saves all the deals in a file called special_feeds_items.txt. 
Information about which kind of attribute is saved for each item can be found in the extractProductInfo function in simple.js

There is an implementation of paginated deal items too that's commented out getPaginatedSpecialFeeds(); . This implementation reports errors while running and items returned could vary from 500 to like 400k items. I don't think it is where we will find the most important opportunities but I added it since it doesn't hurt to investigate

To test, just run node examples/simple.js as usual - once all items are retrieved, run comment out getSpecialFeedsItems() and uncomment getPaginatedSpecialFeeds(); to see how many items get added from that list too.

### CODE REVIEW

File: examples/simple.js [hunk 1]
```diff
@@ -1,17 +1,224 @@
+fs = require('fs');
```
Comment (danfstucky) [inline examples/simple.js:1]: Is this an npm module? It needs to be included in the package.json if it's a dependency

File: examples/simple.js [hunk 2]
```diff
@@ -1,17 +1,224 @@
...
+
+
+/*
+function that serially calls Promise functions and waits for each of them to be resolved
+Reference: https://hackernoon.com/functional-javascript-resolving-promises-sequentially-7aac18c4431e
+*/
+const promiseSerial = cat_array =>
+cat_array.reduce((promise, func) =>
+promise.then(result => func().then(Array.prototype.concat.bind(result))),
+Promise.resolve([]))
```
Comment (danfstucky) [inline examples/simple.js]: Indentation of this block is missing and makes it more confusing to read

File: examples/simple.js [hunk 3]
```diff
@@ -1,17 +1,224 @@
...
+    Gets clearance items available in all categories.
+  */
+  
+  /*
+    after testing for a while, found that feed info for some category_ids return server errors
+    So, omitting them from the category_array we are getting feeds info for.
+    Run writeCategoryIdsToFile() if interested in the names of the categories
+  */
+
+  var invalid_clearance_cat_id_array = ["3920", "976759", "1094765", "6197502", "6197502", "4096", "4104",
```
Comment (danfstucky) [inline examples/simple.js]: I'm not as familiar with this as you, but surely there is a better way than to hard code these right? The clearance items will be changing almost weekly, which means these won't even work a week from now. Would a try/catch block be more appropriate?

File: examples/simple.js [hunk 4]
```diff
@@ -1,17 +1,224 @@
...
+    return;
+  }
+}
+
+
+/*
+function that serially calls Promise functions and waits for each of them to be resolved
+Reference: https://hackernoon.com/functional-javascript-resolving-promises-sequentially-7aac18c4431e
+*/
+const promiseSerial = cat_array =>
```
Comment (danfstucky) [inline examples/simple.js]: I didn't realize node supported ES6 out of the box. In general, it's a good idea to declare all variables that won't be changing as `const` and all mutable variables as `let`. This is the only place you use a const. I'm not gonna make you go and change it everywhere, but wanted to point it out so we can declare variables correctly in the future.

File: examples/simple.js [hunk 5]
```diff
@@ -1,17 +1,224 @@
...
+      var cat_info = category.id+'\n';
+      fs.appendFileSync('category_ids.txt', cat_info);
+    })
+  });
+}
+
+function extractProductInfo(product){
+  var twoThreeDayShippingRate = Object.is(product.twoThreeDayShippingRate, undefined) ? '-1.00' : product.twoThreeDayShippingRate; 
+  var msrp = Object.is(product.msrp, undefined) ? '-1.00' : product.msrp;
+  var product_info=product.itemId+','+product.upc+','+product.name+','+product.brandName+','
```
Comment (danfstucky) [inline examples/simple.js]: Template strings would make this a lot prettier. https://github.com/lukehoban/es6features#template-strings

File: examples/simple.js (hunk 1 above)
Comment (sulaimonlasisi) [reply examples/simple.js:1]: fs is node's in-built filesystem. I don't think it needs to be included anywhere. Ref: https://stackoverflow.com/questions/24594796/node-js-npm-install-fs-error

Comment (danfstucky) [reply examples/simple.js:1]: makes sense

File: examples/simple.js (hunk 3 above)
Comment (sulaimonlasisi) [reply examples/simple.js]: Fixed here: 1e70b04e054b744e3cff3b386ed0487b95d84e5e

File: examples/simple.js [hunk 6]
```diff
@@ -1,17 +1,238 @@
...
+          go through each feed and prepare string combination of feed and cat_id
+          to retrieve information for.
+        */
+          valid_cat_id_array.push(feed+","+category.id); 
+      })
+    });
+
+    // map clearance call over valid_cat_id_array
+    var category_special_feeds_array = valid_cat_id_array.map(feed_and_cat_id => () => getSingleListOfPaginatedItems(feed_and_cat_id));
+    debugger;
```
Comment (danfstucky) [inline examples/simple.js]: remove this

Comment (sulaimonlasisi) [reply examples/simple.js]: I'm improving it at the moment.

Comment (danfstucky) [reply examples/simple.js]: np then. didn't realize

Comment (sulaimonlasisi) [reply examples/simple.js]: Fixed here: 8abf3b3bd3b41232a5185d022ed6d480a4da23e0

File: examples/simple.js (hunk 5 above)
Comment (sulaimonlasisi) [reply examples/simple.js]: Fixed here: 8abf3b3bd3b41232a5185d022ed6d480a4da23e0

**Comment by danfstucky** [comment]
+1

**Comment by sulaimonlasisi** [comment]
Merged at 24154ee33d95856a24571fadfc02cbcc6afbb75b

**Pull request closed without merging**